pkg/controller/v1beta1/inferenceservice/components/explainer.go

```diff
@@ -87,19 +93,158 @@ func (e *Explainer) Reconcile(isvc *v1beta1.InferenceService) (ctrl.Result, erro
+			sRuntime = runtimes[0].Spec
+			isvc.Spec.Explainer.Model.Runtime = &runtimes[0].Name
+
+			// set runtime defaults
+			isvc.SetRuntimeDefaults()
+		}
+		// assign protocol version to inferenceservice based on runtime selected
+		if isvc.Spec.Explainer.Model.ProtocolVersion == nil {
+			protocolVersion := constants.GetProtocolVersionString(
+				constants.ProtocolVersion(
+					v1beta1.GetProtocolVersionPriority(sRuntime.ProtocolVersions),
+				),
+			)
+			isvc.Spec.Explainer.Model.ProtocolVersion = &protocolVersion
+		}
+
+		if len(sRuntime.Containers) == 0 {
+			isvc.Status.UpdateModelTransitionStatus(v1beta1.InvalidSpec, &v1beta1.FailureInfo{
+				Reason:  v1beta1.InvalidExplainerSpec,
+				Message: "No container configuration found in selected serving runtime",
+			})
+			return ctrl.Result{}, errors.New("no container configuration found in selected serving runtime")
+		}
+
+		kserveContainerIdx := -1
+		for i := range sRuntime.Containers {
+			if sRuntime.Containers[i].Name == constants.InferenceServiceContainerName {
+				kserveContainerIdx = i
+				break
+			}
+		}
+		if kserveContainerIdx == -1 {
+			return ctrl.Result{}, errors.New("failed to find kserve-container in ServingRuntime containers")
+		}
+
+		container, err = isvcutils.MergeRuntimeContainers(&sRuntime.Containers[kserveContainerIdx], &isvc.Spec.Explainer.Model.Container)
+		if err != nil {
+			isvc.Status.UpdateModelTransitionStatus(v1beta1.InvalidSpec, &v1beta1.FailureInfo{
+				Reason:  v1beta1.InvalidExplainerSpec,
+				Message: "Failed to get runtime container",
+			})
+			return ctrl.Result{}, errors.Wrapf(err, "failed to get runtime container")
+		}
+
+		mergedPodSpec, err := isvcutils.MergePodSpec(&sRuntime.ServingRuntimePodSpec, &isvc.Spec.Explainer.PodSpec)
+		if err != nil {
+			isvc.Status.UpdateModelTransitionStatus(v1beta1.InvalidSpec, &v1beta1.FailureInfo{
+				Reason:  v1beta1.InvalidExplainerSpec,
+				Message: "Failed to consolidate serving runtime PodSpecs",
+			})
+			return ctrl.Result{}, errors.Wrapf(err, "failed to consolidate serving runtime PodSpecs")
+		}
+
+		// Replace placeholders in runtime container by values from inferenceservice metadata
+		if err = isvcutils.ReplacePlaceholders(container, isvc.ObjectMeta); err != nil {
+			isvc.Status.UpdateModelTransitionStatus(v1beta1.InvalidSpec, &v1beta1.FailureInfo{
+				Reason:  v1beta1.InvalidExplainerSpec,
+				Message: "Failed to replace placeholders in serving runtime Container",
+			})
+			return ctrl.Result{}, errors.Wrapf(err, "failed to replace placeholders in serving runtime Container")
+		}
+
+		// Update image tag if GPU is enabled or runtime version is provided
+		isvcutils.UpdateImageTag(container, isvc.Spec.Explainer.Model.RuntimeVersion, isvc.Spec.Explainer.Model.Runtime)
+
+		podSpec = *mergedPodSpec
+		podSpec.Containers = []v1.Container{
+			*container,
+		}
+		podSpec.Containers = append(podSpec.Containers, sRuntime.Containers[:kserveContainerIdx]...)
```

Comment:
isn't better to check if there are more than 1 container in the runtime.COntainers?

Reply:
I tried to keep it consistent with [the predictor component](https://github.com/TimKleinloog/kserve/blob/pluggable-explainer-runtimes/pkg/controller/v1beta1/inferenceservice/components/predictor.go#L223). Feel free to add a suggestion for an improvement, but I think we should also change it also in the predictor in that case.

Reply:
okay, it is fine :)